umbral/pre.py

```diff
@@ -516,3 +486,27 @@ def decrypt(ciphertext: bytes, capsule: Capsule, decrypting_key: UmbralPrivateKe
         raise UmbralDecryptionError() from e
 
     return cleartext
+
+
+def decrypt_reencrypted(ciphertext: bytes,
+                        capsule: PreparedCapsule,
+                        cfrags: Sequence[CapsuleFrag],
+                        decrypting_key: UmbralPrivateKey,
+                        check_proof: bool = True) -> bytes:
+
+    if not isinstance(ciphertext, bytes) or len(ciphertext) < DEM_NONCE_SIZE:
+        raise ValueError("Input ciphertext must be a bytes object of length >= {}".format(DEM_NONCE_SIZE))
+    elif not isinstance(capsule, PreparedCapsule) or not capsule.verify():
+        raise Capsule.NotValid
+    elif not isinstance(decrypting_key, UmbralPrivateKey):
+        raise TypeError("The decrypting key is not an UmbralPrivateKey")
+
+    encapsulated_key = _open_capsule(capsule, cfrags, decrypting_key, check_proof=check_proof)
+
+    dem = UmbralDEM(encapsulated_key)
+    try:
+        cleartext = dem.decrypt(ciphertext, authenticated_data=bytes(capsule.capsule))
+    except InvalidTag as e:
+        raise UmbralDecryptionError() from e
+
+    return cleartext
```

Comment:
Now that I reread this part, I recall why there was a single decrypt method. It was not only to provide one single entry point for decryption (which is a design choice that you contest in this PR, and I can be fine with the change), but also because the internal logic is identical in both methods, except for the intermediate step (decapsulation). Maybe there's a solution that satisfies both points of view?

Reply:
For this part, why not replace `InvalidTag` with `UmbralDecryptionError` straight in `UmbralDEM.decrypt()`?

Reply:
@cygnusv: Can you help me understand why your design preference has changed regarding a single entry point vs multiple?

@fjarri: In either scenario, raising `UmbralDecryptionError` in the lower frame seems acceptable to me, sure.

Reply:
There is a certain potential barrier for moving `UmbralDecryptionError` to `dem.py`: `GenericUmbralError` will have to be moved somewhere as well, and the applications, which could just import from `.pre` before, will need to import from `.dem` as well. So at the moment I just extracted the DEM operations into a private function in `pre.py`.

Although the right way would be to move ``UmbralDecryptionError` to `dem.py`, since it is a DEM error, and export all the main API from `__init__.py`.